### PR TITLE
Add `clang{,xx}` compilers; build for multiple clang versions

### PR DESCRIPTION
**Pull request by h-vetinari**
This is a continuation of #100, with the main change of not building clang alongside, but elevating it to a separate build, in the spirit of Isuru's [statement](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/100#issuecomment-1815696773):
> I think of this feedstock as Linux compilers

Clang is a full-fledged compiler for linux, so I'd like to treat it like one (by being able to build for several versions). The only overlap between `compiler_vendor == "gcc"` and `"clang"` is that `binutils_{{ cross_target_platform }}` gets built in each job, but will only be uploaded once (due to coinciding hashes). This is _already_ the case before this PR for different GCC versions, so that should be a non-issue IMO.

Rather than building `clang 16 x gcc {11, 12, 13}` (like in #100), I'd prefer to build `clang {16, 17, y} x gcc {base}`. The interaction with the GCC version only comes in through the dependence on `libgcc-devel_{{ cross_target_platform }}`, and I think it's more worthwhile to cover more clang versions, than to enable only old(er) clang, but with all versions of libgcc-devel.

I actually also wanted to tackle https://github.com/conda-forge/clang-compiler-activation-feedstock/issues/118, but [realized](https://github.com/conda-forge/clang-compiler-activation-feedstock/issues/118#issuecomment-1979701833) that we cannot introduce `clang` / `clangxx` outputs without renaming them on the clangdev feedstock first.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -114,13 +162,57 @@ outputs:
...
+        - libgcc-devel_{{ cross_target_platform }} {{ last_symbol_bump }}.*
+        # C++ compiler on linux needs a stdlib
+        - libstdcxx-devel_{{ cross_target_platform }}  {{ last_symbol_bump }}.*
+        # for activation of gcc env vars
+        - {{ pin_subpackage("clang_" ~ cross_target_platform, exact=True) }}
+        # for activation of binutils env vars
+        - {{ pin_subpackage("binutils_" ~ cross_target_platform, exact=True) }}
+        - sysroot_{{ cross_target_platform }}
+        - compiler-rt_{{ cross_target_platform }}
+        - compiler-rt =={{ compiler_version }}.*
```
Comment (h-vetinari) [inline recipe/meta.yaml:182]: Not sure if I'm missing something here, but I think we could leave out a bunch of requirements here, as they're all coming transitively through `clang_{{ cross_target_platform }}`.
```suggestion
        # C++ compiler on linux needs a stdlib
        - libstdcxx-devel_{{ cross_target_platform }}  {{ last_symbol_bump }}.*
        # for activation of gcc env vars
        - {{ pin_subpackage("clang_" ~ cross_target_platform, exact=True) }}
```

**Review by h-vetinari — COMMENTED** [review]
@isuruf @beckermr 
I thought about #100 recently and wanted to see if I can modify it slightly to build for several versions of clang, and it works. 🥳

IMO this setup is cleaner and more extensible - would be interested to hear your thoughts.

It also completes the integration of `clangxx_{{ target }}` already started in #100.

**Comment by beckermr** [comment]
I'll defer to @isuruf. This feedstock has moved beyond me.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {% set build_num = 2 %}
-{% set gcc_major = 13 if ctng_gcc is undefined else ctng_gcc.split(".")[0] %}
+{% set gcc_major = 13 if compiler_version is undefined else compiler_version.split(".")[0] %}
```
Comment (isuruf) [inline recipe/meta.yaml:2]: This value will be 16 or 17 for clang?

Comment (h-vetinari) [reply recipe/meta.yaml:2]: That's true, though all the outputs or lines that use `gcc_major` are skipped for clang builds. I'd be happy to add another condition on `{% if compiler_vendor == "gcc" %}` if you'd like.

Comment (isuruf) [reply recipe/meta.yaml:2]: `min_runtime_version` is a derived from `gcc_major` and is used as a `run_exports` in `clang`.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -114,13 +162,57 @@ outputs:
...
+      commands:
+        - ${CXX} ${CXXFLAGS} -Wall tests/aligned_alloc.cpp -c -o cpp_aligned.o
+        - ${CXX} ${LDFLAGS} cpp_aligned.o -o cpp_aligned    # [target_platform == cross_target_platform]
+        - ./cpp_aligned                                     # [target_platform == cross_target_platform]
+        - test -z "${CONDA_BUILD_SYSROOT+x}" && echo "CONDA_BUILD_SYSROOT is not set" && exit 1
+        - test -d ${CONDA_BUILD_SYSROOT} || exit 1
+    run_exports:
+      strong:
+        # This should be a transitive dependency, but conda-build doesn't support those
+        - libstdcxx-ng >={{ last_symbol_bump }}
```
Comment (isuruf) [inline recipe/meta.yaml:198]: This should correspond to the `libstdcxx_devel` version used here.

File: recipe/meta.yaml (hunk 2 above)
Comment (h-vetinari) [reply recipe/meta.yaml:2]: I changed the run-export to use `last_symbol_bump` (==12). It seemed to me like a reasonable base GCC version - we don't need the newest GCC when using clang, and we have to use the `last_symbol_bump` as the run-export lower bound in any case for anything below.

File: recipe/meta.yaml (hunk 3 above)
Comment (h-vetinari) [reply recipe/meta.yaml:198]: It does, both are fixed to `last_symbol_bump` ==12

**Comment by h-vetinari** [comment]
Obsoleted by #102

**Pull request closed without merging**